File: Examples/test-suite/perl5/reference_global_vars_runme.pl
use strict;
use warnings;
use Test::More tests => 19;
BEGIN { use_ok('reference_global_vars') }
require_ok('reference_global_vars');

# adapted from ../python/reference_global_vars_runme.py

BEGIN {
  # don't try this at home kids... sneaking an import of all symbols
  # from reference_global_vars to main because my fingers are getting
  # sore from qualifying all these names. ;)
  %{*::} = (%{*reference_global_vars::}, %{*::});
}

is(getconstTC()->{num}, 33, 'const class reference variable');

$var_bool = createref_bool(0);
is(value_bool($var_bool), '', 'bool false');

$var_bool = createref_bool(1);
is(value_bool($var_bool), 1, 'bool true');

$var_char = createref_char('w');
is(value_char($var_char), 'w', 'char');

$var_unsigned_char = createref_unsigned_char(10);
is(value_unsigned_char($var_unsigned_char), 10, 'uchar');

$var_signed_char = createref_signed_char(10);
is(value_signed_char($var_signed_char), 10, 'schar');

$var_short = createref_short(10);
is(value_short($var_short), 10, 'short');

$var_unsigned_short = createref_unsigned_short(10);
is(value_unsigned_short($var_unsigned_short), 10, 'ushort');

$var_int = createref_int(10);
is(value_int($var_int), 10, 'int');

$var_unsigned_int = createref_unsigned_int(10);
is(value_unsigned_int($var_unsigned_int), 10, 'uint');

$var_long = createref_long(10);
is(value_long($var_long), 10, 'long');

$var_unsigned_long = createref_unsigned_long(10);
is(value_unsigned_long($var_unsigned_long), 10, 'ulong');

SKIP: {
  no warnings 'portable';
  skip "64 bit int support", 1 unless eval { pack 'q', 1 };
  my $ll = '6FFFFFFFFFFFFFF8';
  $var_long_long = createref_long_long(hex $ll);
  is(value_long_long($var_long_long), hex $ll, 'long long');

  my $ull = 'FFFFFFF2FFFFFFF0';
  $var_unsigned_long_long = createref_unsigned_long_long(hex $ull);
  is(value_unsigned_long_long($var_unsigned_long_long), hex $ull, 'ulong long');
}

$var_float = createref_float(10.5);
is(value_float($var_float), 10.5, 'float');

$var_double = createref_double(10.5);
is(value_double($var_double), 10.5, 'double');

$var_TestClass = createref_TestClass(TestClass->new(20));
is(value_TestClass($var_TestClass)->{num}, 20, 'class reference variable');

File: Examples/test-suite/ocaml/reference_global_vars_runme.ml
open Swig
open Reference_global_vars

let _ =
  let tc = _getconstTC '() in
  assert (tc -> "[num]" () as int = 33);

  let _ = _var_bool (_createref_bool (C_bool false)) in
  assert (_value_bool (_var_bool '()) as bool = false);

  let _ = _var_bool (_createref_bool (C_bool true)) in
  assert (_value_bool (_var_bool '()) as bool = true);

  let _ = _var_char (_createref_char (C_char 'w')) in
  assert (_value_char (_var_char '()) as char = 'w');

  let _ = _var_unsigned_char (_createref_unsigned_char (C_uchar 'w')) in
  assert (_value_unsigned_char (_var_unsigned_char '()) as char = 'w');

  let _ = _var_signed_char (_createref_signed_char (C_uchar 'w')) in
  assert (_value_signed_char (_var_signed_char '()) as char = 'w');

  let _ = _var_short (_createref_short (C_short 10)) in
  assert (_value_short (_var_short '()) as int = 10);

  let _ = _var_unsigned_short (_createref_unsigned_short (C_ushort 10)) in
  assert (_value_unsigned_short (_var_unsigned_short '()) as int = 10);

  let _ = _var_int (_createref_int (C_int 10)) in
  assert (_value_int (_var_int '()) as int = 10);

  let _ = _var_unsigned_int (_createref_unsigned_int (C_int 10)) in
  assert (_value_unsigned_int (_var_unsigned_int '()) as int = 10);

  let _ = _var_long (_createref_long (C_int64 10L)) in
  assert (_value_long (_var_long '()) as int = 10);

  let _ = _var_unsigned_long (_createref_unsigned_long (C_int64 10L)) in
  assert (_value_unsigned_long (_var_unsigned_long '()) as int = 10);

  let _ = _var_long_long (_createref_long_long (C_int64 0x6FFFFFFFFFFFFFF8L)) in
  assert (_value_long_long (_var_long_long '()) = C_int64 0x6FFFFFFFFFFFFFF8L);

  let _ = _var_unsigned_long_long (_createref_unsigned_long_long (C_int64 0x6FFFFFFFFFFFFFF8L)) in
  assert (_value_unsigned_long_long (_var_unsigned_long_long '()) = C_int64 0x6FFFFFFFFFFFFFF8L);

  let _ = _var_float (_createref_float (C_float 10.)) in
  assert (_value_float (_var_float '()) as float = 10.);

  let _ = _var_double (_createref_double (C_double 10.)) in
  assert (_value_double (_var_double '()) as float = 10.);

  let tc = new_TestClass '(20) in
  let _ = _var_TestClass (_createref_TestClass (tc)) in
  let tc = _value_TestClass (_var_TestClass '()) in
  assert (tc -> "[num]" () as int = 20);
;;
